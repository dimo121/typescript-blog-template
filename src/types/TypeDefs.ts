import { CognitoUser } from "@aws-amplify/auth";


export interface NewUserInput {
    username?: string;
    email: string;
    password: string;
}

export interface NewEntryInput {
    title:string;
    content:string;
    user:string;
    blog_id:string;
}

export interface NewBlogInput {
    title:string;
    content:string;
    user:string;
}

export interface User {
    username: string;
    user: CognitoUser;
}

export interface UserAttribute {
    Name: string;
    Value: string;
}

type Item = {
    id: string,
    title: string,
    content: string,
    createdAt: string,
    user: string,
    owner?: {
        username: string
    }
}

export interface Blog extends Item{
    entries?: Entry[];
}

export interface Entry extends Item{
    blog_id: number;
}

export type Search = 'Title'|'Content';
