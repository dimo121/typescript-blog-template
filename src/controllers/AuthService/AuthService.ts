import { User, UserAttribute } from '../../types/TypeDefs';
import Amplify, { Auth } from 'aws-amplify';
import { config } from '../../config';
import { CognitoUser } from '@aws-amplify/auth';
// import { Credentials } from 'aws-sdk/lib/credentials';
// import AWS from 'aws-sdk';


Amplify.configure({
    Auth: {
        mandatorySignIn: false,
        region: config.REGION,
        identityPoolId: config.IDENTITY_POOL_ID,
        userPoolId: config.USER_POOL_ID,
        userPoolWebClientId: config.APP_CLIENT_ID,
        authenticationFlowType: 'USER_SRP_AUTH'
    }
})

export class AuthService {

    public async login(email: string, password:string): Promise<User|undefined>{
        try{
            const user = await Auth.signIn(email, password) as CognitoUser;

            // await this.getAWSTemporaryCreds(user);

            return {
                username: user.getUsername(),
                user
            };
        } catch (error){
            return undefined;
        }
    }

    public async getUserAttributes(user: User):Promise<UserAttribute[]>{
        const result : UserAttribute[] = []
        const attributes = await Auth.userAttributes(user.user);
        result.push(...attributes);
        return result;
    }

    public async signUp(username:string,password:string,email:string): Promise<User|undefined> {
        try {
            const { user } = await Auth.signUp({
                username,
                password,
                attributes: {
                    email
                }
            });
        
            return {
                username: user.getUsername(),
                user
            };
        } catch (error) {
            console.log('Error signing up:', error);
            return undefined;
        }
    }

    public async confirmSignUp(username:string, code:string):Promise<boolean> {
        try{
            await Auth.confirmSignUp(username, code);
            return true;
        } catch(error){
            console.log('Error confirming sign up', error)
            return false;
        }
    }

    public async signOut() {
        try{
            await Auth.signOut();
        }catch(error){
            console.log('Error signing out: ', error)
        }
    }

    // private async getAWSTemporaryCreds(user: CognitoUser){
    //     const cognitoIdentityPool = `cognito-idp.${config.REGION}.amazonaws.com/${config.USER_POOL_ID}`; 
    //     AWS.config.credentials = new AWS.CognitoIdentityCredentials({
    //         IdentityPoolId: config.IDENTITY_POOL_ID,
    //         Logins: {
    //             [cognitoIdentityPool]: user.getSignInUserSession()!.getIdToken().getJwtToken()
    //         }
    //     }, {
    //         region: config.REGION
    //     });
    //     await this.refreshCredentials();
    // }

    
    // private async refreshCredentials(): Promise<void>{
    //     return new Promise((resolve, reject)=>{
    //         (AWS.config.credentials as Credentials).refresh(err =>{
    //             if (err) {
    //                 reject(err)
    //             } else {
    //                 resolve()
    //             }
    //         })
    //     })
    // }

}
