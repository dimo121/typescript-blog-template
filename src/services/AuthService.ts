import { User, UserAttribute } from '../types/TypeDefs';
import { Auth } from 'aws-amplify';
import Amplify from 'aws-amplify';
import { config } from './config';
import { CognitoUser } from '@aws-amplify/auth';
//import * as AWS from 'aws-sdk';
//import { Credentials } from 'aws-sdk/lib/credentials';


Amplify.configure({
    Auth: {
        mandatorySignIn: false,
        region: config.REGION,
        userPoolId: config.USER_POOL_ID,
        userPoolWebClientId: config.APP_CLIENT_ID,
        //identityPoolId: config.IDENTITY_POOL_ID,
        authenticationFlowType: 'USER_SRP_AUTH'
    }
})

export class AuthService {

    public async login(email: string, password:string): Promise<User|undefined>{
        try{
            const user = await Auth.signIn(email, password) as CognitoUser;
            return {
                username: user.getUsername(),
                user
            };
        } catch (error){
            return undefined;
        }
    }

    public async getUserAttributes(user: User):Promise<UserAttribute[]>{
        const result : UserAttribute[] = []
        const attributes = await Auth.userAttributes(user.user);
        result.push(...attributes);
        return result;
    }

    public async signUp(username:string,password:string,email:string): Promise<User|undefined> {
        try {
            const { user } = await Auth.signUp({
                username,
                password,
                attributes: {
                    email
                }
            });
            console.log(user);
            return {
                username: user.getUsername(),
                user
            };
        } catch (error) {
            console.log('error signing up:', error);
            return undefined;
        }
    }

    // public async getAWSTemporaryCreds(user:CognitoUser){
    //     const cognitoUserPool = `cognito-idp.${config.REGION}.amazonaws.com/${config.USER_POOL_ID}`;
    //     AWS.config.credentials = new AWS.CognitoIdentityCredentials({
    //         IdentityPoolId: config.IDENTITY_POOL_ID,
    //         Logins: {
    //             [cognitoUserPool]: user.getSignInUserSession()!.getIdToken().getJwtToken()
    //         }
    //     }, {
    //         region: config.REGION
    //     });
    //     await this.refreshCredentials()
    // }

    // private async refreshCredentials(): Promise<void>{
    //     return new Promise((resolve,reject) => {
    //         (AWS.config.credentials as Credentials).refresh(err => {
    //             if(err){
    //                 reject(err)
    //             } else{
    //                 resolve()
    //             }
    //         })
    //     })
    // }

}
