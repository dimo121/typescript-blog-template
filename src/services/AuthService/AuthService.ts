import { User, UserAttribute } from '../../types/TypeDefs';
import { Auth } from 'aws-amplify';
import Amplify from 'aws-amplify';
import { config } from '../config';
import { CognitoUser } from '@aws-amplify/auth';


Amplify.configure({
    Auth: {
        mandatorySignIn: false,
        region: config.REGION,
        userPoolId: config.USER_POOL_ID,
        userPoolWebClientId: config.APP_CLIENT_ID,
        authenticationFlowType: 'USER_SRP_AUTH'
    }
})

export class AuthService {

    public async login(email: string, password:string): Promise<User|undefined>{
        try{
            const user = await Auth.signIn(email, password) as CognitoUser;
            return {
                username: user.getUsername(),
                user
            };
        } catch (error){
            return undefined;
        }
    }

    public async getUserAttributes(user: User):Promise<UserAttribute[]>{
        const result : UserAttribute[] = []
        const attributes = await Auth.userAttributes(user.user);
        result.push(...attributes);
        return result;
    }

    public async signUp(username:string,password:string,email:string): Promise<User|undefined> {
        try {
            const { user } = await Auth.signUp({
                username,
                password,
                attributes: {
                    email
                }
            });
            console.log(user);
            return {
                username: user.getUsername(),
                user
            };
        } catch (error) {
            console.log('error signing up:', error);
            return undefined;
        }
    }

    public async confirmSignUp(username:string, code:string) {
        try{
            await Auth.confirmSignUp(username, code);
        } catch(error){
            console.log('Error confirming sign up', error)
        }
    }

    public async signOut() {
        try{
            await Auth.signOut();
        }catch(error){
            console.log('Error signing out: ', error)
        }
    }

}
