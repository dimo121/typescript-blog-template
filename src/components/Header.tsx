import React, { useState } from 'react';
import img from './react-logo.png';
import Modal from './Modal';
import { AuthService } from '../controllers/AuthService/AuthService';
import { User, NewUserInput } from '../types/TypeDefs';
import { NavLink } from 'react-router-dom';


interface IHeaderProps {
    authService: AuthService;
    user?: User;
    setUser: (authUser:User|undefined) => void;
}

const Header:React.FC<IHeaderProps> = (props) => {
    
    const { authService, user, setUser } = props;

    const [modalOpen,setModal] = useState<boolean>(false);
    const [isActive, setActive] =  useState<boolean>(false);


    const openMenu = ():void => {
        document.body.classList.add('nav-is-active');
        setActive(true);
    }

    const closeMenu = ():void => {
        document.body.classList.remove('nav-is-active');
        setActive(false);
    }

    const onSignInClick = ():void => {
        if(user){
            setUser(undefined);
        }else{
            setModal(true);
            document.body.classList.remove('nav-is-active')
        }
    }
    
    const clearModal = ():void => {
        setModal(false);
    };
  
    const onSubmission = async (user:NewUserInput): Promise<void> => {
        if(!user.username){
            const authUser = await authService.login(user.email, user.password);
            setUser(authUser);
        }
        else{
            await authService.signUp(user.username,user.password,user.email);
        }
    }


    return (
        <header className="s6-hc">
            <div className="s6-hc__title">
                <span>
                    <h1><img  src={img} className='s6-hc__logo' alt="React logo" />React.JS blog</h1>
                </span>
                <button className="s6-hc__toggle" onClick={isActive ? closeMenu: openMenu}>
                    <span className="hamburger" data-testid="hamburger">
                        <span className="hamburger-inner"></span>
                    </span>
                </button>
            </div>
            <div className="s6-hc__nc">
                <ul>
                    <li>
                        <NavLink exact onClick={closeMenu} activeClassName="s6-hc__link--active" to="/" tabIndex={1}>
                            Home
                        </NavLink>
                    </li>
                    <li>
                        <NavLink onClick={closeMenu} activeClassName="s6-hc__link--active" to="/dashboard" tabIndex={2}>
                            Dashboard
                        </NavLink>
                    </li>
                    <li>
                        <NavLink onClick={closeMenu} activeClassName="s6-hc__link--active" to="/createblog" tabIndex={3}>
                            Create blog
                        </NavLink>
                    </li>
                    <li>
                        <NavLink onClick={closeMenu} activeClassName="s6-hc__link--active" to="/myblogs" tabIndex={4}>
                            My blogs
                        </NavLink>
                    </li>
                    <li>
                        <div className="s6-hc__signin">
                            <h3>{user ? `Signed in as : ${user.username}` : ''}</h3>
                            <button className="main-button" onClick={onSignInClick} tabIndex={5}>
                                {user ? "Sign Out" : "Sign In"}
                            </button>
                        </div>
                    </li>
                </ul>
            </div>
            <Modal
                modalOpen={modalOpen}
                clearModal={clearModal}
                onSubmission={onSubmission}
                authService={authService}
            />
        </header>
    )
}

export default Header;
