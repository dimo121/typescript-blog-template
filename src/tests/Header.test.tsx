import { render, screen } from '@testing-library/react';
import Header from '../components/Header';
import { AuthService } from '../services/AuthService/AuthService';
import { User } from '../types/TypeDefs';
import { Router } from 'react-router-dom';
import history from '../utils/history';


const authService = new AuthService();

function setUser(user:User|undefined){
    
  console.log('Logged in user is: ', user?.username);
};


test('renders header title React.JS blog', () => {

  render(<Router history={history}>
    <Header setUser={setUser} authService={authService}/>
    </Router>);
  const linkElement = screen.getByText(/React.JS blog/i);
  expect(linkElement).toBeInTheDocument();

});


test('should take a snapshot of the header', () => {
  const { container } = render(<Router history={history}>
    <Header setUser={setUser} authService={authService}/>
    </Router>);

  expect(container).toMatchSnapshot();
})

test('should render hamburger for mobile interface only', () => {
    
    render(<Router history={history}>
            <Header setUser={setUser} authService={authService}/>
          </Router>);

    global.innerWidth = 520;
    global.innerHeight = 750;
    global.dispatchEvent(new Event('resize'));


    expect(window.innerWidth).toEqual(520);
    expect(window.innerHeight).toEqual(750);
    
    expect(screen.getByTestId('hamburger')).toBeInTheDocument();
    //expect(button).toHaveAttribute('class','nav-toggle');
    //still renders with display : none
})