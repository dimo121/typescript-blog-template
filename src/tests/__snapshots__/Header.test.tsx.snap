// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot of the header 1`] = `
<div>
  <header
    class="header-container"
  >
    <div
      class="header-title"
    >
      <span>
        <h1>
          <img
            alt="React logo"
            class="header-logo"
            src="react-logo.png"
          />
          React.JS blog
        </h1>
      </span>
      <button
        class="nav-toggle"
      >
        <span
          class="hamburger"
          data-testid="hamburger"
        >
          <span
            class="hamburger-inner"
          />
        </span>
      </button>
    </div>
    <div
      class="nav-container"
    >
      <ul>
        <li>
          <a
            aria-current="page"
            class="nav-links active"
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            class="nav-links"
            href="/dashboard"
          >
            Dashboard
          </a>
        </li>
        <li>
          <a
            class="nav-links"
            href="/createblog"
          >
            Create blog
          </a>
        </li>
        <li>
          <a
            class="nav-links"
            href="/myblogs"
          >
            My blogs
          </a>
        </li>
        <li>
          <p
            class="nav-signin"
          />
          <button
            class="main-button"
          >
            Sign In
          </button>
        </li>
      </ul>
    </div>
  </header>
</div>
`;
